### PR TITLE
Implement a custom event formatter for javascript logging

### PR DESCRIPTION
**Pull request by richvdh**
Somehow `fmt::pretty` manages to be both overcomplicated and not flexible enough.

The driver here is that I want to put the event "fields" on a separate line to the message.

### CODE REVIEW

**Comment by richvdh** [comment]
I think the coverage metrics must be wrong here. There is definitely a test that runs this code and checks the result. I think it is run [here](https://github.com/matrix-org/matrix-rust-sdk/actions/runs/6921199553/job/18826538783?pr=2856#step:9:43) although the test name seems to have sprouted an `_outer` suffix from somewhere.

**Comment by jplatte** [comment]
We don't (and probably can't) collect coverage on WASM-only code. You can change the `cfg(target_arch = "wasm32")` attribute on the `js_tracing` module to `cfg(all(target_arch = "wasm32", not(tarpaulin_include)))` to exclude it from coverage reporting.

**Comment by richvdh** [comment]
> You can change the cfg(target_arch = "wasm32") attribute on the js_tracing module to cfg(all(target_arch = "wasm32", not(tarpaulin_include))) to exclude it from coverage reporting.

Is that the right thing to do? If we're unable to measure code coverage on a portion of our code, it's probably better that our metrics reflect that, even if it means overriding the coverage check on PRs that touch that code?

**Comment by jplatte** [comment]
I don't think I should be answering that question 😄

**Comment by richvdh** [comment]
In this case, it  should be possible to test `JsEventFormatter` outside wasm, since it doesn't rely on the js console. But that would mean bringing `tracing-subscriber` as a (dev?) dependency, and building *parts* of the `js_tracing` module outside wasm, and suddenly it's a whole bunch of work just to satisfy a coverage metric.

**Comment by poljar** [comment]
> Is that the right thing to do? If we're unable to measure code coverage on a portion of our code, it's probably better that our metrics reflect that, even if it means overriding the coverage check on PRs that touch that code?

The other argument can be made as well, if we know that the code is tested, why do we want code coverage to tell us it isn't?

I think it's fine to disable coverage collection in this case, because:

1. The code is tested.
2. The bot will pester us needlessly that we should add a test for it when we don't need to.
3. People looking at the coverage reports will see a file with 0% code coverage and will try to write test for it, when in fact there are tests.

Out of those, no 3 sounds the worst to me. Disabling the gathering with a comment explaining that we have tests sounds sensible to avoid this situation. What exactly do we gain from not disabling it? 

On the other hand, I have trouble understanding why this piece of code is even in the SDK repo.

**Comment by richvdh** [comment]
> > Is that the right thing to do? If we're unable to measure code coverage on a portion of our code, it's probably better that our metrics reflect that, even if it means overriding the coverage check on PRs that touch that code?
> 
> The other argument can be made as well, if we know that the code is tested, why do we want code coverage to tell us it isn't?
> 
> I think it's fine to disable coverage collection in this case, because:
> 
>     1. The code is tested.
>     2. The bot will pester us needlessly that we should add a test for it when we don't need to.
>     3. People looking at the coverage reports will see a file with 0% code coverage and will try to write test for it, when in fact there are tests.
> 
> Out of those, no 3 sounds the worst to me. Disabling the gathering with a comment explaining that we have tests sounds sensible to avoid this situation. What exactly do we gain from not disabling it?

Those are fair arguments, particularly because in this case it is, in fact, tested. The counterpoint is if we start to accumulate (even more) cruft in this module which is *not* tested, and the metric won't pick it up. But let's agree not to do that.

> On the other hand, I have trouble understanding why this piece of code is even in the SDK repo.

Fair question. The reason that `MakeJsLogWriter` is in this repo is basically so that we can do logging from within tests of other parts of the SDK that have to be done in the wasm environment (eg the indexeddb store). And I dumped `JsEventFormatter` here because they went together in my head, but that's not actually right and I could certainly move it to the wasm bindings repo if you prefer.

**Comment by poljar** [comment]
> And I dumped JsEventFormatter here because they went together in my head, but that's not actually right and I could certainly move it to the wasm bindings repo if you prefer.

It sounds like you'd like to put this into a separate crate, the IndexedDB crate and the WASM bindings can then use this new crate. Something like https://github.com/jquesada2016/tracing_subscriber_wasm or https://github.com/old-storyai/tracing-wasm might also be interesting for the IndexedDB/tests case.

In any case, I'm fine with disabling the gathering of the coverage and moving forward with this PR.

**Comment by richvdh** [comment]
> Something like https://github.com/jquesada2016/tracing_subscriber_wasm or https://github.com/old-storyai/tracing-wasm might also be interesting for the IndexedDB/tests case.

Hrm yes. I hadn't found those before. One thing I wanted to achieve with `MakeJsLogWriter` was the ability to pass in a "Logger" object other than the console, and `tracing_subscriber_wasm` lacks that ability. Still, it would make a lot of sense to use `tracing_subscriber_wasm` in the tests and then we could put this `js_logging` module into the WASM bindings of the rust-sdk where it probably belongs. (Or, as you say, in a completely separate crate). I'll bear this in mind for the future.

I also like `tracing_wasm`'s idea of wiring up spans to the JS `Performance` API. It looks like it might be really helpful in tracking down performance problems. Something to explore in future.

> In any case, I'm fine with disabling the gathering of the coverage and moving forward with this PR.

Ok thanks; that is what I have done for now.

File: crates/matrix-sdk-common/src/js_tracing.rs [hunk 1]
```diff
@@ -199,9 +202,130 @@ fn write_message_to_console(level: Level, message: &JsValue) {
     };
 }
 
+/// An implementation of [`FormatEvent`] which formats events in a sensible way
+/// for sending events to the JS console.
+#[derive(Debug, Default)]
+pub struct JsEventFormatter {}
```
Comment (jplatte) [inline crates/matrix-sdk-common/src/js_tracing.rs:208]:
```suggestion
#[non_exhaustive]
pub struct JsEventFormatter {}
```

If we want to enforce that `::new()` is the way to create an instance.

File: crates/matrix-sdk-common/src/js_tracing.rs [hunk 2]
```diff
@@ -199,9 +202,130 @@ fn write_message_to_console(level: Level, message: &JsValue) {
...
+        }
+
+        // write the other fields
+        let mut v = JsFieldVisitor::new(writer.by_ref());
+        event.record(&mut v);
+
+        if let Some(file) = meta.file() {
+            write!(writer, "\n    at {}", file)?;
+            if let Some(line) = meta.line() {
+                write!(writer, ":{}", line)?;
```
Comment (jplatte) [inline crates/matrix-sdk-common/src/js_tracing.rs:244]:
```suggestion
            write!(writer, "\n    at {file}")?;
            if let Some(line) = meta.line() {
                write!(writer, ":{line}")?;
```

File: crates/matrix-sdk-common/src/js_tracing.rs [hunk 3]
```diff
@@ -199,9 +202,130 @@ fn write_message_to_console(level: Level, message: &JsValue) {
...
+            }
+        }
+
+        let span = event.parent().and_then(|id| ctx.span(id)).or_else(|| ctx.lookup_current());
+        let scope = span.into_iter().flat_map(|span| span.scope());
+        for span in scope {
+            let meta = span.metadata();
+            write!(writer, "\n    in {}::{}", meta.target(), meta.name(),)?;
```
Comment (jplatte) [inline crates/matrix-sdk-common/src/js_tracing.rs:252]:
```suggestion
            write!(writer, "\n    in {}::{}", meta.target(), meta.name())?;
```

File: crates/matrix-sdk-common/src/js_tracing.rs [hunk 4]
```diff
@@ -199,9 +202,130 @@ fn write_message_to_console(level: Level, message: &JsValue) {
...
+
+            let ext = span.extensions();
+            let fields = &ext
+                .get::<FormattedFields<N>>()
+                .expect("Unable to find FormattedFields in extensions; this is a bug");
+            if !fields.is_empty() {
+                write!(writer, " with {}", fields)?;
```
Comment (jplatte) [inline crates/matrix-sdk-common/src/js_tracing.rs:259]:
```suggestion
                write!(writer, " with {fields}")?;
```

File: crates/matrix-sdk-common/src/js_tracing.rs [hunk 5]
```diff
@@ -199,9 +202,130 @@ fn write_message_to_console(level: Level, message: &JsValue) {
...
+/// for the event.
+#[derive(Debug, Default)]
+struct FindMessageVisitor {
+    message: Option<String>,
+}
+
+impl tracing::field::Visit for FindMessageVisitor {
+    fn record_debug(&mut self, field: &Field, value: &dyn Debug) {
+        if field.name() == "message" {
+            self.message = Some(format!("{:?}", value));
```
Comment (jplatte) [inline crates/matrix-sdk-common/src/js_tracing.rs:277]:
```suggestion
            self.message = Some(format!("{value:?}"));
```

File: crates/matrix-sdk-common/src/js_tracing.rs [hunk 6]
```diff
@@ -199,9 +202,130 @@ fn write_message_to_console(level: Level, message: &JsValue) {
...
+        // If this is the first field since the message, make a new line. Otherwise,
+        // just print a space.
+        if self.is_empty {
+            self.is_empty = false;
+            write!(self.writer, "\n    ")?;
+        } else {
+            write!(self.writer, " ")?;
+        }
+
+        write!(self.writer, "{}={:?}", name, value)
```
Comment (jplatte) [inline crates/matrix-sdk-common/src/js_tracing.rs:305]:
```suggestion
        write!(self.writer, "{name}={value:?}")
```

**Review by jplatte — APPROVED** [review]

**Pull request merged**